### PR TITLE
Paginate list datum

### PR DESCRIPTION
**Pull request by sjezewski**
This PR adds:

- a `full` flag to `ListFile`, which means the `children` field is populated in the return value
- the `paginate` and `page` flags on the `ListDatum` API call
- a bunch of tests

For now, it seemed simplest to return an `io.EOF` if a page is requested that is past the bounds of the resulting datums. This is because we plan on detecting errors on the UI. This may require updating as the UI's requirements become more clear.

### CODE REVIEW

File: src/server/pachyderm_test.go [hunk 1]
```diff
@@ -3686,14 +3686,146 @@ func TestPipelineWithStats(t *testing.T) {
...
+			EnableStats: true,
+			ParallelismSpec: &pps.ParallelismSpec{
+				Constant: 4,
+			},
+		})
+
+	_, err = c.FlushCommit([]*pfs.Commit{commit1}, nil)
+	require.NoError(t, err)
+
+	// Without this sleep, I get no results from list-job
```
Comment (msteffen) [inline src/server/pachyderm_test.go]: Maybe open an issue for this, if there isn't one already?

File: src/server/pps/cmds/cmds.go [hunk 2]
```diff
@@ -235,6 +237,8 @@ $ pachctl list-job -p foo bar/YYY
 		}),
 	}
 	rawFlag(listDatum)
+	listDatum.Flags().BoolVar(&paginate, "paginate", false, "Get a subset of the results")
+	listDatum.Flags().Int64Var(&page, "page", 0, "Specify the page of results to send")
```
Comment (sjezewski) [inline src/server/pps/cmds/cmds.go]: I'm not sure we can do that ... bcz non initialized ints are 0 right? (And the pages are zero indexed) So I can't tell the difference between non initialized page and page=0 (the first page).

But ... this did make me realize it needs to be an unsigned int

Comment (jdoliner) [inline src/server/pps/cmds/cmds.go]: That `0` there is a default value so you can just change that to negative `-1` and you'll be able to see if it was specified or not. Also I don't really think unsigned ints are worth it, even though the value shouldn't ever be negative, you just wind up having to do a million casts.

File: src/server/pps/server/api_server.go [hunk 3]
```diff
@@ -511,62 +511,79 @@ func (a *apiServer) ListDatum(ctx context.Context, request *pps.ListDatumRequest
 		}
 		return datumHash, nil
 	}
+	for _, fileInfo := range allFileInfos.FileInfo {
+		_, err := pathToDatumHash(fileInfo.File.Path)
```
Comment (sjezewski) [inline src/server/pps/server/api_server.go]: Hmm ... not sure what you mean. I need to remove some of the file infos before I sort them, and that is done whether or not the response is paginated.

File: src/server/pps/cmds/cmds.go (hunk 2 above)
Comment (sjezewski) [inline src/server/pps/cmds/cmds.go]: Ah, I just saw this. [I actually made the change and it wans't any additional casts really](https://github.com/pachyderm/pachyderm/commit/a8e5e199a50dd18aac808c60bc163762036bb926)

But if you want int64, I can do that. I will have to add some error checking though (right now if you specify a negative page, I believe we'll panic). Using `uint64` seems like an easier way to deal to validate the input consistently.

Comment (sjezewski) [inline src/server/pps/cmds/cmds.go]: Ok, after offline disc going w int64's ... but also we're using `pageSize` (a new param) w a non zero value as the measure if we should paginate.

File: src/server/pps/server/api_server.go (hunk 3 above)
Comment (sjezewski) [inline src/server/pps/server/api_server.go]: Ah, you're asking to use the composite if statement style ... ok done

**Pull request merged**